File: crates/scred-detector/src/simd_core.rs
//! SIMD Core: Fast prefix search with portable SIMD for charset validation
//! 
//! Strategy (matching Zig optimization patterns):
//! - Use memchr for single-byte matching (2-4x faster than manual loop)
//! - Process in chunks (16-32 bytes) for batch operations
//! - Charset validation using lookup tables (not regex)
//! - SIMD for candidate filtering (portable via slice iterators)
//! - Fallback to memchr for multi-byte prefixes

use memchr::memchr;

/// Portable SIMD-style charset validation via lookup table
/// This is faster than scanning every byte
#[derive(Clone, Copy)]
pub struct CharsetLut {
    pub table: [bool; 256],  // ← Now public for SIMD access
}

impl CharsetLut {
    /// Create a lookup table for a charset
    pub fn new(charset: &[u8]) -> Self {
        let mut table = [false; 256];
        for &byte in charset {
            table[byte as usize] = true;
        }
        CharsetLut { table }
    }

    #[inline]
    pub fn contains(&self, byte: u8) -> bool {
        self.table[byte as usize]
    }

    /// Scan data for end of token (first byte NOT in charset)
    /// Zig optimization: scanForTokenEnd32 but portable
    /// Optimized: Uses SIMD when available, falls back to scalar
    #[inline]
    pub fn scan_token_end(&self, data: &[u8], start: usize) -> usize {
        crate::simd_charset::scan_token_end_fast(data, self, start)
    }
}

/// Find first occurrence of prefix in data
/// Uses memchr for first byte, then validates full prefix
#[inline]
pub fn find_first_prefix(data: &[u8], prefix: &[u8]) -> Option<usize> {
    if data.is_empty() || prefix.is_empty() {
        return if prefix.is_empty() { Some(0) } else { None };
    }

    if prefix.len() > data.len() {
        return None;
    }

    let first_byte = prefix[0];

    // Fast path: single-byte prefix
    if prefix.len() == 1 {
        return memchr(first_byte, data);
    }

    // Multi-byte prefix: use memchr to find candidates, then validate
    let mut search_start = 0;
    while let Some(pos) = memchr(first_byte, &data[search_start..]) {
        let absolute_pos = search_start + pos;
        
        // Check if we have enough bytes for full prefix
        if absolute_pos + prefix.len() <= data.len() {
            // Validate full prefix at this position
            if &data[absolute_pos..absolute_pos + prefix.len()] == prefix {
                return Some(absolute_pos);
            }
        }
        
        // Move search forward
        search_start = absolute_pos + 1;
    }

    None
}

/// Find all occurrences of prefix in data (up to max_matches)
pub fn find_all_prefixes(data: &[u8], prefix: &[u8], max_matches: usize) -> Vec<usize> {
    let mut matches = Vec::with_capacity(max_matches.min(10));

    if data.is_empty() || prefix.is_empty() || prefix.len() > data.len() {
        return matches;
    }

    if max_matches == 0 {
        return matches;
    }

    let first_byte = prefix[0];
    let mut search_start = 0;

    while let Some(pos) = memchr(first_byte, &data[search_start..]) {
        if matches.len() >= max_matches {
            break;
        }

        let absolute_pos = search_start + pos;

        // Check if we have enough bytes
        if absolute_pos + prefix.len() <= data.len() {
            // Validate full prefix
            if &data[absolute_pos..absolute_pos + prefix.len()] == prefix {
                matches.push(absolute_pos);
            }
        }

        // Continue search from next byte
        search_start = absolute_pos + 1;
    }

    matches
}

/// Validate token length and charset (Zig optimization pattern)
pub fn validate_token(
    data: &[u8],
    start: usize,
    min_len: usize,
    max_len: usize,
    charset: &CharsetLut,
) -> Option<(usize, usize)> {
    if start >= data.len() {
        return None;
    }

    // Scan token end using charset LUT (O(n) where n is token len)
    let token_len = charset.scan_token_end(data, start);

    // Check constraints
    if token_len < min_len {
        return None;
    }

    if max_len > 0 && token_len > max_len {
        return None;
    }

    Some((start, start + token_len))
}

File: crates/scred-pattern-detector/src/simd_core.zig
//! SIMD Core: Real vectorized operations for pattern matching
//!
//! Strategy: Use Zig @Vector for batch processing
//! - 16-byte chunks: Compare prefix patterns in parallel
//! - 32-byte chunks: Extend token scanning
//! - Fallback to scalar for <16 bytes
//!
//! Expected improvement: 2-4x speedup on prefix scanning

const std = @import("std");

/// Find all positions where any of the prefixes start in a 16-byte chunk
/// Returns a vector of 16 bools indicating matches
pub fn findPrefixesIn16Bytes(
    chunk: [16]u8,
    prefixes: []const []const u8,
) @Vector(16, bool) {
    var matches: @Vector(16, bool) = @splat(false);

    // For each prefix, check if it could start at each byte position
    for (prefixes) |prefix| {
        if (prefix.len == 0) continue;
        if (prefix.len > 16) continue; // Skip long prefixes for chunk matching

        const first_byte: u8 = prefix[0];
        const fb_vec: @Vector(16, u8) = @splat(first_byte);
        const chunk_vec: @Vector(16, u8) = chunk;

        // Check where first byte matches
        const first_byte_matches: @Vector(16, bool) = chunk_vec == fb_vec;

        // Mark as potential match (full validation happens in scalar path)
        matches = matches or first_byte_matches;
    }

    return matches;
}

/// Find first matching position in data using 16-byte SIMD chunks
/// Falls back to scalar search for final bytes
pub fn findFirstPrefix(
    data: []const u8,
    prefix: []const u8,
) ?usize {
    if (data.len < prefix.len) return null;
    if (prefix.len == 0) return 0;

    const first_byte = prefix[0];
    var pos: usize = 0;

    // Process 16-byte chunks
    while (pos + 16 <= data.len) : (pos += 16) {
        const chunk: [16]u8 = data[pos..][0..16].*;
        const chunk_vec: @Vector(16, u8) = chunk;
        const fb_vec: @Vector(16, u8) = @splat(first_byte);

        // Check if first byte matches anywhere in chunk
        const matches: @Vector(16, bool) = chunk_vec == fb_vec;

        // Check each match position
        for (0..16) |i| {
            if (matches[i]) {
                const candidate_pos = pos + i;
                if (candidate_pos + prefix.len <= data.len) {
                    if (std.mem.eql(u8, data[candidate_pos .. candidate_pos + prefix.len], prefix)) {
                        return candidate_pos;
                    }
                }
            }
        }
    }

    // Handle remaining bytes (<16)
    if (pos < data.len) {
        const remaining = data[pos..];
        if (std.mem.indexOf(u8, remaining, prefix)) |rel_pos| {
            return pos + rel_pos;
        }
    }

    return null;
}

/// Find all matching positions (up to max_matches) using SIMD
pub fn findAllPrefixes(
    allocator: std.mem.Allocator,
    data: []const u8,
    prefix: []const u8,
    max_matches: usize,
) !std.ArrayList(usize) {
    var matches = std.ArrayList(usize).init(allocator);

    if (data.len < prefix.len or prefix.len == 0) {
        return matches;
    }

    const first_byte = prefix[0];
    var pos: usize = 0;

    // Process 16-byte chunks
    while (pos + 16 <= data.len and matches.items.len < max_matches) : (pos += 16) {
        const chunk: [16]u8 = data[pos..][0..16].*;
        const chunk_vec: @Vector(16, u8) = chunk;
        const fb_vec: @Vector(16, u8) = @splat(first_byte);

        const match_vec: @Vector(16, bool) = chunk_vec == fb_vec;

        // Check each match position
        for (0..16) |i| {
            if (match_vec[i]) {
                const candidate_pos = pos + i;
                if (candidate_pos + prefix.len <= data.len) {
                    if (std.mem.eql(u8, data[candidate_pos .. candidate_pos + prefix.len], prefix)) {
                        try matches.append(candidate_pos);
                        if (matches.items.len >= max_matches) break;
                    }
                }
            }
        }
    }

    // Handle remaining bytes (<16)
    if (pos < data.len and matches.items.len < max_matches) {
        var remaining_pos = pos;
        while (remaining_pos + prefix.len <= data.len) : (remaining_pos += 1) {
            if (std.mem.eql(u8, data[remaining_pos .. remaining_pos + prefix.len], prefix)) {
                try matches.append(remaining_pos);
                if (matches.items.len >= max_matches) break;
            }
        }
    }

    return matches;
}

/// Check if any byte in a 32-byte chunk matches any in the charset
/// Used for token boundary detection
pub fn scanForTokenEnd32(
    data: []const u8,
    start: usize,
    charset: []const u8,
) usize {
    if (start >= data.len) return 0;

    var pos = start;
    var len: usize = 0;

    // Process 32-byte chunks for efficiency
    const chunk_size: usize = 32;
    while (pos + chunk_size <= data.len) : (pos += chunk_size) {
        const chunk = data[pos .. pos + chunk_size];
        var chunk_len: usize = 0;

        for (chunk) |byte| {
            // Check if byte is IN charset
            var found = false;
            for (charset) |cs_byte| {
                if (byte == cs_byte) {
                    found = true;
                    break;
                }
            }

            if (found) {
                chunk_len += 1;
            } else {
                // End of token
                return len + chunk_len;
            }
        }

        len += chunk_len;
    }

    // Handle remaining bytes
    while (pos < data.len) : (pos += 1) {
        const byte = data[pos];
        var found = false;
        for (charset) |cs_byte| {
            if (byte == cs_byte) {
                found = true;
                break;
            }
        }

        if (found) {
            len += 1;
        } else {
            break;
        }
    }

    return len;
}

/// Batch filter: given multiple candidates, check which actually match the full prefix
/// Returns count of true matches
pub fn filterCandidates(
    data: []const u8,
    candidates: []const usize,
    prefix: []const u8,
    max_valid: usize,
) usize {
    var valid_count: usize = 0;

    for (candidates) |candidate_pos| {
        if (valid_count >= max_valid) break;
        if (candidate_pos + prefix.len > data.len) continue;

        if (std.mem.eql(u8, data[candidate_pos .. candidate_pos + prefix.len], prefix)) {
            valid_count += 1;
        }
    }

    return valid_count;
}

// ============================================================================
// Tests
// ============================================================================

test "findFirstPrefix finds position" {
    const data = "hello AKIAIOSFODNN7EXAMPLE world";
    const prefix = "AKIA";

    const pos = findFirstPrefix(data, prefix);
    try std.testing.expectEqual(@as(?usize, 6), pos);
}

test "findFirstPrefix returns null when not found" {
    const data = "hello world";
    const prefix = "NOTHERE";

    const pos = findFirstPrefix(data, prefix);
    try std.testing.expectEqual(@as(?usize, null), pos);
}

test "findFirstPrefix works at start" {
    const data = "AKIAIOSFODNN7EXAMPLE";
    const prefix = "AKIA";

    const pos = findFirstPrefix(data, prefix);
    try std.testing.expectEqual(@as(?usize, 0), pos);
}

test "findAllPrefixes finds multiple positions" {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const data = "AKIAIOSFODNN7EXAMPLE some text AKIAIOSFODNN7EXAMPLE end";
    const prefix = "AKIA";

    const matches = try findAllPrefixes(allocator, data, prefix, 10);
    defer matches.deinit();

    try std.testing.expectEqual(@as(usize, 2), matches.items.len);
    try std.testing.expectEqual(@as(usize, 0), matches.items[0]);
    try std.testing.expectEqual(@as(usize, 36), matches.items[1]);
}

test "scanForTokenEnd32 detects boundaries" {
    const data = "sk_prod_abc123XYZ, next";
    const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_";

    const len = scanForTokenEnd32(data, 0, charset);
    try std.testing.expectEqual(@as(usize, 13), len);
}
